src/hyp_map.rs

```diff
@@ -165,21 +211,42 @@ impl HypMapPopulatedRegion {
             .zip(pages.base().iter_from())
             .take(page_count as usize)
         {
-            // Safe because these pages are mapped into user mode and will not be accessed in
-            // supervisor mode.
+            // Safety: it is not an alias because these are user mode mappings and these specific
+            // mappings cannot be accessed (without special functions) from supervisor mode.
             unsafe {
                 mapper.map_addr(virt, phys, self.pte_fields).unwrap();
             }
         }
+        // If writable user region, return a U-mode reset area.
+        if self.resettable {
+            // Guaranteed not to overflow by contruction of `page_count`.
+            let mapped_size = page_count * PageSize::Size4k as u64;
+            Some(UmodeResetArea {
+                paddr: pages.base(),
+                mapped_size,
+                size: self.size,
+                data: self.data,
+            })
+        } else {
+            None
+        }
     }
 }
 
 /// A page table that contains hypervisor mappings.
 pub struct HypPageTable {
     inner: FirstStagePageTable<Sv48>,
+    umode_reset: ArrayVec<UmodeResetArea, MAX_HYPMAP_UMODE_REGIONS>,
```

Comment:
AFAICT the reason these are stored in the per-CPU `HypPageTable` is so that we can use the supervisor-level mappings from the identity map to access the U-mode memory, and those addresses will obviously differ between CPUs.

Any reason we can't use the (user) virtual addresses instead? The commit message says that using the supervisor mapping is faster, but is it really that big of a difference? We should only need to flip `sstatus.SUM` and then copying should "just work", e.g.

```
CSR.sstatus.modify(sstatus::sum.val(1));
core::ptr::copy(data.as_ptr(), dest_va, len);
CSR.sstatus.modify(sstatus::sum.val(0));
```

We know the mappings exist here, so we don't need to do the PF fixup tricks like we do when copying to/from guest memory.

And if you do end up using the user VA, there's no need to keep these reset areas around per-CPU; keeping the U-mode `ElfMap` in a single (immutable) global should be sufficient.

Reply:
Okay. This is how I implemented it originally. Then I had a look at how guest mem is copied and realised that we're extremely precise on the PC that might cause an exception. So I went and implemented copy_to/from_user (and a bzero) in the same way. Then realised that in the lack of optimization that would be a huge performance penalty. And came up with this.

What I did was consuming the `ElfMap` in the `Umode::init`, as parsing it everytime we reset is a bit pointless.

Will do.

Reply:
Changed. See `umode_mem.rs`.

Reply:
Reset implementation has been removed. `umode_mem.rs` as well.